File: apps/android/app/src/main/java/com/rayact/engine/RayactPlatformRegistry.kt
package com.rayact.engine

import android.content.Context
import android.app.Activity
import android.content.Intent
import android.view.View
import org.json.JSONObject
import java.util.concurrent.ConcurrentHashMap
import java.util.concurrent.CountDownLatch
import java.util.concurrent.TimeUnit
import java.util.concurrent.atomic.AtomicBoolean
import java.lang.ref.WeakReference

fun interface RayactPlatformModule {
    fun invoke(method: String, payload: ByteArray, completion: (Result<ByteArray>) -> Unit)
}

class RayactPlatformViewContext(
    val context: Context,
    val nodeId: Int,
    val initialProperties: Map<String, Any?>,
    val emit: (String) -> Unit,
)

interface RayactPlatformViewController {
    val view: View
    fun setProperties(properties: Map<String, Any?>) {}
    fun dispose() {}
}

fun interface RayactPlatformViewFactory {
    fun create(context: RayactPlatformViewContext): RayactPlatformViewController
}

interface RayactPlatformModuleRegistration {
    fun register(context: Context, registry: RayactPlatformRegistry)
}

class RayactPlatformRegistry {
    private val modules = ConcurrentHashMap<String, RayactPlatformModule>()
    private val viewFactories = ConcurrentHashMap<String, RayactPlatformViewFactory>()

    fun registerModule(name: String, module: RayactPlatformModule) {
        require(name.isNotBlank()) { "Platform module name must not be blank" }
        check(modules.putIfAbsent(name, module) == null) { "Platform module '$name' already registered" }
    }

    fun registerViewFactory(kind: String, factory: RayactPlatformViewFactory) {
        require(kind.isNotBlank()) { "Platform view kind must not be blank" }
        check(viewFactories.putIfAbsent(kind, factory) == null) { "Platform view '$kind' already registered" }
    }

    fun invoke(name: String, method: String, payload: ByteArray, completion: (Result<ByteArray>) -> Unit) {
        val module = modules[name]
        if (module == null) {
            completion(Result.failure(IllegalStateException("Platform module '$name' is not installed")))
            return
        }
        module.invoke(method, payload, completion)
    }

    fun createView(kind: String, context: RayactPlatformViewContext): RayactPlatformViewController? =
        viewFactories[kind]?.create(context)

    fun hasModule(name: String): Boolean = modules.containsKey(name)
    fun hasViewFactory(kind: String): Boolean = viewFactories.containsKey(kind)

    fun emitActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
        val payload = JSONObject()
            .put("requestCode", requestCode)
            .put("resultCode", resultCode)
            .put("uri", data?.dataString ?: JSONObject.NULL)
            .toString()
            .toByteArray(Charsets.UTF_8)
        modules.values.forEach { module ->
            module.invoke("__activityResult", payload) { }
        }
    }

    companion object {
        val shared = RayactPlatformRegistry()
        private val initialized = AtomicBoolean(false)
        @Volatile private var foregroundActivity = WeakReference<Activity>(null)
        private const val CALL_TIMEOUT_SECONDS = 15L

        fun initialize(context: Context) {
            if (context is Activity) attachActivity(context)
            if (!initialized.compareAndSet(false, true)) return
            registerRayactNativeTextInput(shared)
            runCatching {
                val generated = Class.forName("com.rayact.generated.RayactGeneratedModules")
                generated.getMethod(
                    "register",
                    Context::class.java,
                    RayactPlatformRegistry::class.java,
                ).invoke(null, context.applicationContext, shared)
            }.onFailure {
                initialized.set(false)
                if (it !is ClassNotFoundException) throw it
            }
        }

        fun attachActivity(activity: Activity) {
            foregroundActivity = WeakReference(activity)
        }

        fun currentActivity(): Activity? =
            foregroundActivity.get()?.takeUnless { it.isFinishing || it.isDestroyed }

        @JvmStatic
        fun platformCallFromNative(name: String, method: String, payloadJson: String?): String {
            val latch = CountDownLatch(1)
            var response = "{\"ok\":false,\"error\":\"Platform module call did not complete\"}"
            shared.invoke(name, method, (payloadJson ?: "null").toByteArray(Charsets.UTF_8)) { result ->
                response = result.fold(
                    onSuccess = {
                        val json = it.toString(Charsets.UTF_8).ifBlank { "null" }
                        "{\"ok\":true,\"value\":$json}"
                    },
                    onFailure = {
                        "{\"ok\":false,\"error\":${JSONObject.quote(it.message ?: "Platform module call failed")}}"
                    },
                )
                latch.countDown()
            }
            if (!latch.await(CALL_TIMEOUT_SECONDS, TimeUnit.SECONDS)) {
                return "{\"ok\":false,\"error\":\"Platform module call timed out\"}"
            }
            return response
        }
    }
}

File: apps/ios/RayactPlatformRegistry.swift
import UIKit
import Darwin

typealias RayactPlatformCompletion = (Result<Data, Error>) -> Void
typealias RayactPlatformModule = (_ method: String, _ payload: Data, _ completion: @escaping RayactPlatformCompletion) -> Void

final class RayactPlatformViewContext {
    let nodeId: Int
    let initialProperties: [String: Any]
    let emit: (String) -> Void

    init(nodeId: Int, initialProperties: [String: Any], emit: @escaping (String) -> Void) {
        self.nodeId = nodeId
        self.initialProperties = initialProperties
        self.emit = emit
    }
}

protocol RayactPlatformViewController: AnyObject {
    var view: UIView { get }
    func setProperties(_ properties: [String: Any?])
    func dispose()
}

extension RayactPlatformViewController {
    func setProperties(_ properties: [String: Any?]) {}
    func dispose() {}
}

typealias RayactPlatformViewFactory = (_ context: RayactPlatformViewContext) -> RayactPlatformViewController

protocol RayactPlatformModuleRegistration {
    func register(with registry: RayactPlatformRegistry)
}

final class RayactPlatformRegistry {
    static let shared = RayactPlatformRegistry()
    private static let initializationLock = NSLock()
    private static var initialized = false
    private static let responseLock = NSLock()
    private static var responseStorage: [CChar] = [0]

    private var modules: [String: RayactPlatformModule] = [:]
    private var viewFactories: [String: RayactPlatformViewFactory] = [:]
    private let lock = NSLock()

    func registerModule(_ name: String, invoke: @escaping RayactPlatformModule) {
        precondition(!name.isEmpty, "Platform module name must not be empty")
        lock.lock()
        defer { lock.unlock() }
        precondition(modules[name] == nil, "Platform module '\(name)' already registered")
        modules[name] = invoke
    }

    func registerViewFactory(_ kind: String, factory: @escaping RayactPlatformViewFactory) {
        precondition(!kind.isEmpty, "Platform view kind must not be empty")
        lock.lock()
        defer { lock.unlock() }
        precondition(viewFactories[kind] == nil, "Platform view '\(kind)' already registered")
        viewFactories[kind] = factory
    }

    func invoke(_ name: String, method: String, payload: Data, completion: @escaping RayactPlatformCompletion) {
        lock.lock()
        let module = modules[name]
        lock.unlock()
        guard let module else {
            completion(.failure(NSError(
                domain: "RayactPlatformModule",
                code: 1,
                userInfo: [NSLocalizedDescriptionKey: "Platform module '\(name)' is not installed"]
            )))
            return
        }
        module(method, payload, completion)
    }

    func makeViewController(
        _ kind: String,
        nodeId: Int,
        properties: [String: Any],
        emit: @escaping (String) -> Void
    ) -> RayactPlatformViewController? {
        lock.lock()
        let factory = viewFactories[kind]
        lock.unlock()
        return factory?(RayactPlatformViewContext(
            nodeId: nodeId,
            initialProperties: properties,
            emit: emit
        ))
    }

    func hasModule(_ name: String) -> Bool {
        lock.lock()
        defer { lock.unlock() }
        return modules[name] != nil
    }

    static func foregroundViewController() -> UIViewController? {
        dispatchPrecondition(condition: .onQueue(.main))
        let root = UIApplication.shared.connectedScenes
            .compactMap { $0 as? UIWindowScene }
            .flatMap(\.windows)
            .first(where: \.isKeyWindow)?
            .rootViewController
        return topViewController(from: root)
    }

    private static func topViewController(from controller: UIViewController?) -> UIViewController? {
        if let presented = controller?.presentedViewController {
            return topViewController(from: presented)
        }
        if let navigation = controller as? UINavigationController {
            return topViewController(from: navigation.visibleViewController)
        }
        if let tabs = controller as? UITabBarController {
            return topViewController(from: tabs.selectedViewController)
        }
        return controller
    }

    func emitSystemEvent(_ type: String, payload: [String: Any] = [:]) {
        lock.lock()
        let installedModules = Array(modules.values)
        lock.unlock()
        var event = payload
        event["type"] = type
        guard let data = try? JSONSerialization.data(withJSONObject: event) else { return }
        for module in installedModules {
            module("__systemEvent", data) { _ in }
        }
    }

    static func initialize() {
        initializationLock.lock()
        defer { initializationLock.unlock() }
        guard !initialized else { return }
        initialized = true
        registerRayactNativeTextInput(with: shared)
        guard let symbol = dlsym(
            UnsafeMutableRawPointer(bitPattern: -2),
            "RayactRegisterGeneratedModules"
        ) else {
            initialized = false
            return
        }
        typealias Register = @convention(c) () -> Void
        unsafeBitCast(symbol, to: Register.self)()
    }

    static let platformCallCallback: RayactNativeBridge.PlatformCallFn = {
        modulePointer,
        methodPointer,
        payloadPointer in
        let module = modulePointer.map(String.init(cString:)) ?? ""
        let method = methodPointer.map(String.init(cString:)) ?? ""
        let payload = Data((payloadPointer.map(String.init(cString:)) ?? "null").utf8)
        let semaphore = DispatchSemaphore(value: 0)
        var response = Data(#"{"ok":false,"error":"Platform module call did not complete"}"#.utf8)
        RayactPlatformRegistry.shared.invoke(module, method: method, payload: payload) { result in
            switch result {
            case .success(let value):
                let json = String(data: value, encoding: .utf8).flatMap { $0.isEmpty ? nil : $0 } ?? "null"
                response = Data("{\"ok\":true,\"value\":\(json)}".utf8)
            case .failure(let error):
                response = (try? JSONSerialization.data(withJSONObject: [
                    "ok": false,
                    "error": error.localizedDescription,
                ])) ?? Data(#"{"ok":false,"error":"Platform module call failed"}"#.utf8)
            }
            semaphore.signal()
        }
        if semaphore.wait(timeout: .now() + 15) == .timedOut {
            response = Data(#"{"ok":false,"error":"Platform module call timed out"}"#.utf8)
        }
        responseLock.lock()
        responseStorage = response.map { CChar(bitPattern: $0) } + [0]
        let pointer = responseStorage.withUnsafeBufferPointer { $0.baseAddress }
        responseLock.unlock()
        return pointer
    }
}
